### PR TITLE
add twin topic support

### PR DESCRIPTION
**Pull request by danewalton**
Adds the topic get functionality for twin (pub and sub). Parsing responses will come later after #513.

### CODE REVIEW

File: sdk/iot/hub/inc/az_iot_hub_client.h [hunk 1]
```diff
@@ -460,18 +460,13 @@ AZ_NODISCARD az_result az_iot_hub_client_twin_get_publish_topic_get(
  *
  * @param[in] client The #az_iot_hub_client to use for this call.
  * @param[in] request_id The request id.
- * @param[in] if_match_version Can be either "*" to overwrite or the twin version to limit the
```
Comment (CIPop) [inline sdk/iot/hub/inc/az_iot_hub_client.h]: Are these not part of the topic? Can they only be accessed via the payload?

File: sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+static void test_az_iot_hub_client_twin_response_subscribe_topic_filter_get_null_client_fails(
+    void** state)
+{
+  (void)state;
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_response_topic_filter) - 1];
+  az_span test_span = az_span_init(test_span_buffer, 0, _az_COUNTOF(test_span_buffer));
+
+  assert_precondition_checked(
```
Comment (CIPop) [inline sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c]: This name is confusing - maybe assert_precondition_failed? (Not part of this PR though it's the first time I see it.)

File: sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+static void test_az_iot_hub_client_twin_response_subscribe_topic_filter_get_succeed(void** state)
+{
+  (void)state;
+
+  az_iot_hub_client client;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, NULL), AZ_OK);
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_response_topic_filter) - 1];
```
Comment (CIPop) [inline sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c]: I'd actually initialize this with fixed space (128, 256) to:

1. demonstrate the way we need app-developers to think about the buffers.
2. check to see the buffer is not overrun

File: sdk/iot/hub/inc/az_iot_hub_client.h (hunk 1 above)
Comment (danewalton) [reply sdk/iot/hub/inc/az_iot_hub_client.h]: I confirmed with Ewerton this is not for MQTT or AMQP. When I did a search in the current SDK it only came up in HTTP.

File: sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c (hunk 2 above)
Comment (danewalton) [reply sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c]: Yea we could have a separate PR for changing if we want for sure.

File: sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c (hunk 3 above)
Comment (danewalton) [reply sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c]: I kinda see your point about demonstrating use case but I'm not so sure that unit tests is the place where people should go to understand how to use API's.
As for the buffer overrun part I can see that too.
I would like to issue another PR that checks the span length for all tests. @hihigupt had brought that up previously and I think it's an sdk test improvement. Is it okay to push to another PR focusing solely on that?
